plugin/kubernetes/containerBase.go

```diff
@@ -0,0 +1,108 @@
+package kubernetes
+
+import (
+	"context"
+	"io"
+
+	"github.com/puppetlabs/wash/activity"
+	corev1 "k8s.io/api/core/v1"
+	k8s "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/remotecommand"
+)
+
+// A general purpose container object that implements executing commands.
+// If the pod contains only a single container, the container object may be omitted.
+type containerBase struct {
+	client    *k8s.Clientset
+	config    *rest.Config
+	pod       *corev1.Pod
+	container *corev1.Container
+}
+
+// Create an executor to run a command. If you want synchronous results, call executor.Stream(). For
+// asynchronous results call the `async` method.
+func (c *containerBase) exec(ctx context.Context, cmd string, args []string, opts remotecommand.StreamOptions) (remotecommand.Executor, error) {
+	execRequest := c.client.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Name(c.pod.Name).
+		Namespace(c.pod.Namespace).
+		SubResource("exec").
+		Param("command", cmd)
+
+	if c.container != nil {
+		execRequest = execRequest.Param("container", c.container.Name)
+	}
+
+	if opts.Stdout != nil {
+		execRequest = execRequest.Param("stdout", "true")
+	}
+
+	if opts.Stderr != nil {
+		execRequest = execRequest.Param("stderr", "true")
+	}
+
+	if opts.Stdin != nil || opts.Tty {
+		execRequest = execRequest.Param("stdin", "true")
+	}
+
+	if opts.Tty {
+		execRequest = execRequest.Param("tty", "true")
+	}
+
+	for _, arg := range args {
+		execRequest = execRequest.Param("command", arg)
+	}
+
+	return remotecommand.NewSPDYExecutor(c.config, "POST", execRequest.URL())
+}
+
+func (c *containerBase) async(ctx context.Context, errHandler func(error), e remotecommand.Executor, opts remotecommand.StreamOptions) (cleanup func()) {
```

Comment:
Is this basically `asyncStream`? If so, I think creating a wrapper type around `remotecommand.Executor` would help. Then code reads as `executor.Stream()`, `executor.AsyncStream(errHandler) (stopFunc)` (the type would store the context, Stream options, etc.).